### PR TITLE
Implement opset v10 changes for Slice operator

### PR DESCRIPTION
**Pull request by hariharans29**
Slice has had a spec change (https://github.com/onnx/onnx/commit/ce332628c0de863ab9847272b37a650076e0b23e) and along with the slightly different interface, it accommodates arbitrary "stepping" along the specified axes. This changes implements that functionality.

### CODE REVIEW

**Comment by hariharans29** [comment]
CC: @snnn @skottmckay @pranavsharma  for review comments please

File: onnxruntime/core/providers/cpu/tensor/slice.cc [hunk 1]
```diff
@@ -79,59 +107,160 @@ Status SliceBase::PrepareForCompute(const std::vector<int64_t>& raw_starts,
...
+  if (axes.size() == 0) {
+    // axes are omitted, they are set to[0, ..., ndim - 1]
+    axes.resize(starts.size());
+    std::iota(axes.begin(), axes.end(), 0);
+  }
+
+  // Iterate through the provided axes and override the start/end/steps ranges
+  std::unordered_set<int64_t> unique_axes;
+  const auto& dimension_count = input_dimensions.size();
+  for (size_t axesIndex = 0; axesIndex < axes.size(); axesIndex++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/slice.cc]:
```suggestion
  for (size_t axesIndex = 0, end = axes.size(); axesIndex < end; axesIndex++) {
```

File: onnxruntime/core/providers/cpu/tensor/slice.cc [hunk 2]
```diff
@@ -79,59 +107,160 @@ Status SliceBase::PrepareForCompute(const std::vector<int64_t>& raw_starts,
...
+                                    const std::vector<int64_t>& raw_ends,
+                                    const std::vector<int64_t>& raw_axes,
+                                    const std::vector<int64_t>& raw_steps,
+                                    const std::vector<int64_t>& input_dimensions,
+                                    std::vector<int64_t>& starts,
+                                    std::vector<int64_t>& steps,
+                                    std::vector<int64_t>& output_dims) const {
+  // Initialize axes to the provided axes attribute or to the default sequence
+  std::vector<int64_t> axes(raw_axes);
+  if (axes.size() == 0) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/slice.cc]:
```suggestion
  if (!axes.empty()) {
```

File: onnxruntime/core/providers/cpu/tensor/utils.h [hunk 3]
```diff
@@ -119,47 +119,69 @@ struct ExtentAxisCounters {
...
+
+    size_t pitch = 1;
+    for (size_t i = size(); i-- > 0;) {
       auto prevPitch = pitch;
       pitch *= dims[i];
-      operator[](i) = pitch - prevPitch * extents[i];
+      
+      // assume step == 1 if not present
+      int64_t steps_i_minus_1 = 1;
+      if (i > 0 && static_cast<ptrdiff_t>(i)-1 < steps.size())
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/utils.h]: Can steps.size() be called outside the for loop?

File: onnxruntime/core/providers/cpu/tensor/slice.cc (hunk 2 above)
Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/slice.cc]: changed to axes.empty() as the previous line had `axes.size() == 0`

File: onnxruntime/core/providers/cpu/tensor/slice.cc [hunk 4]
```diff
@@ -82,59 +111,160 @@ Status SliceBase::PrepareForCompute(const std::vector<int64_t>& raw_starts,
   }
 
   // Iterate through the provided axes and override the start/end ranges
+  std::unordered_set<int64_t> unique_axes;
   const auto& dimension_count = input_dimensions.size();
   for (size_t axesIndex = 0; axesIndex < axes.size(); axesIndex++) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/tensor/slice.cc]: We don't use camel case for variable names.

Comment (hariharans29) [reply onnxruntime/core/providers/cpu/tensor/slice.cc]: This is how it was originally in the file. I make sure not to introduce camel-casing for newly created variables.

**Pull request merged**